[improve][cli]Support for pulsar-shell persistent modes

sandeep-mst opened the pull request:
<!--
### Contribution Checklist
  
  - PR title format should be *[type][component] summary*. For details, see *[Guideline - Pulsar PR Naming Convention](https://pulsar.apache.org/contribute/develop-semantic-title/)*. 

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.
-->

<!-- Either this PR fixes an issue, -->

Fixes [21194](https://github.com/apache/pulsar/issues/21194) of apache/pulsar

### Motivation
<!-- Explain here the context, and why you're making that change. What is the problem you're trying to solve. -->
Pulsar shell doesn't support persistent mode for admin, client, config. If we need to run a lot of admin commands for example, we would need to prefix admin each time. This will provide an alternative approach.

### Modifications

<!-- Describe the modifications you've done. -->
Modified so that `admin`, `client`, `config` commands individually will launch their own mode and prompt message also gets updated. `exit` or `quit` command will exit from the mode to the default mode.
In the admin mode, we can run command `topics` instead of `admin topics`.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is already covered by existing tests, such as *(please describe tests)*.

### Does this pull request potentially affect one of the following parts:

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

### Matching PR in forked repository

PR in forked repository: <!-- ENTER URL HERE -->
https://github.com/cognitree/pulsar/pull/6
<!--
After opening this PR, the build in apache/pulsar will fail and instructions will
be provided for opening a PR in the PR author's forked repository.

apache/pulsar pull requests should be first tested in your own fork since the 
apache/pulsar CI based on GitHub Actions has constrained resources and quota.
GitHub Actions provides separate quota for pull requests that are executed in 
a forked repository.

The tests will be run in the forked repository until all PR review comments have
been handled, the tests pass and the PR is approved by a reviewer.
-->

lhotari commented on pulsar-client-tools/src/main/java/org/apache/pulsar/shell/PulsarShell.java:
```diff
@@ -414,18 +441,36 @@ public List<String> readCommand() {
             execState = ExecState.RUNNING;
-            final String line = words.stream().collect(Collectors.joining(" "));
-            if (StringUtils.isBlank(line)) {
+            final String inputLine = words.stream().collect(Collectors.joining(" "));
```
perhaps `final` could be dropped so that the a single variable `line` could be used.

sandeep-mst replied:
Updated it

lhotari commented on pulsar-client-tools/src/main/java/org/apache/pulsar/shell/PulsarShell.java:
```diff
@@ -414,18 +441,36 @@ public List<String> readCommand() {
+            if (isQuitCommand(inputLine)) {
+                if (shellMode != ShellMode.DEFAULT){
+                    output(String.format("Exiting from %s shell mode", shellMode.command), terminal);
+                    promptMessage = promptMessage.substring(0, promptMessage.lastIndexOf(shellMode.command) - 1);
```
It might be better to keep the current shell mode out of the promptMessage and build the `prompt` from `promptMessage` and `shellMode.command` each time there are changes.

sandeep-mst replied:
yes, updated it.

lhotari commented on pulsar-client-tools/src/main/java/org/apache/pulsar/shell/PulsarShell.java:
```diff
@@ -414,18 +441,36 @@ public List<String> readCommand() {
             }
-            if (isQuitCommand(line)) {
+            if (isQuitCommand(inputLine)) {
+                if (shellMode != ShellMode.DEFAULT){
```
Whitespace
```suggestion
                if (shellMode != ShellMode.DEFAULT) {
```

lhotari commented on pulsar-client-tools/src/main/java/org/apache/pulsar/shell/PulsarShell.java:
```diff
@@ -414,18 +441,36 @@ public List<String> readCommand() {
                 continue;
             }
+            ShellMode newShellMode = ShellMode.valueOfCommand(line.toLowerCase(Locale.ROOT));
+            if (newShellMode != null){
```
whitespace
```suggestion
            if (newShellMode != null) {
```

lhotari commented on pulsar-client-tools/src/main/java/org/apache/pulsar/shell/PulsarShell.java:
```diff
@@ -414,18 +441,36 @@ public List<String> readCommand() {
+            ShellMode newShellMode = ShellMode.valueOfCommand(line.toLowerCase(Locale.ROOT));
+            if (newShellMode != null){
+                shellMode = newShellMode;
+                promptMessage = promptMessage.concat(" ").concat(shellMode.command);
+                prompt = createPrompt(promptMessage);
```
See the previous comment about not modifying promptMessage. In Java, the compiler will optimize String concatenation. You can simply use `+` to concatenate Strings.
```suggestion
                prompt = createPrompt(promptMessage + " " + shellMode.command));
```

lhotari reviewed (commented):
Good work @sandeep-mst . A few minor review comments.

nicoloboschi reviewed (changes requested):
I like this idea. 

1. I think we should register the new commands like other commands in `registerProviders(..)` so automatic autocompletion will work for these modes too. 

2. I don't understand the title, what do you mean by persistent modes ? I think we need a better name for this feature. 

3. We should add docs, at least in the help (might be mitigated by 1.)

lhotari commented on pulsar-client-tools/src/main/java/org/apache/pulsar/shell/PulsarShell.java:
```diff
@@ -414,18 +441,34 @@ public List<String> readCommand() {
                 shellCommander.usage(System.out);
                 continue;
             }
+            ShellMode newShellMode = ShellMode.valueOfCommand(line.toLowerCase(Locale.ROOT));
+            if (newShellMode != null) {
```
checking for this makes only sense when `shellMode == ShellMode.DEFAULT`. Switching the shell mode should only be possible at the top level.

sandeep-mst replied:
Yes, I updated it. Now switching will only be possible at the top level.

lhotari commented:
> I like this idea.
> 
> 1. I think we should register the new commands like other commands in `registerProviders(..)` so automatic autocompletion will work for these modes too.
> 2. I don't understand the title, what do you mean by persistent modes ? I think we need a better name for this feature.
> 3. We should add docs, at least in the help (might be mitigated by 1.)

@sandeep-mst Have you considered addressing @nicoloboschi's review feedback?
For number 2 point, calling it a "subcontext" would be perhaps better unless that concept carries already a different meaning.

sandeep-mst commented:
> > I like this idea.
> > 
> > 1. I think we should register the new commands like other commands in `registerProviders(..)` so automatic autocompletion will work for these modes too.
> > 2. I don't understand the title, what do you mean by persistent modes ? I think we need a better name for this feature.
> > 3. We should add docs, at least in the help (might be mitigated by 1.)
> 
> @sandeep-mst Have you considered addressing @nicoloboschi's review feedback? For number 2 point, calling it a "subcontext" would be perhaps better unless that concept carries already a different meaning.
> 
> For example, [windows netsh has the concept of contexts and subcontexts](https://learn.microsoft.com/en-us/windows-server/networking/technologies/netsh/netsh-contexts#netsh-subcontexts) (top level is "context", which could have "subcontexts").

Hi @lhotari, 

1. I have addressed the @nicoloboschi review feedback for supporting auto-completion for shell modes for `admin` and `client`. I removed `config` from the provided shell modes.
> I tried to implement `config` shell mode even though it was not requested in the issue. I was facing lot of issues with it because `config` is stateful and implementation might require major changes to support auto-completion.

2. On the PR naming, context is already being used by [window-functions-context](https://pulsar.apache.org/docs/3.3.x/window-functions-context/) of pulsar. We could still use it. Please, let me know your thoughts.
